src/middlewared/middlewared/plugins/filesystem_/acl_linux.py

```diff
@@ -12,82 +13,102 @@ class FilesystemService(Service, ACLBase):
 
     @private
     def acltool(self, path, action, uid, gid, options):
+
+        flags = "-r"
+        flags += "x" if options.get('traverse') else ""
+        flags += "C" if options.get('do_chmod') else ""
+        flags += "P" if options.get('posixacl') else ""
+
         acltool = subprocess.run([
-            '/usr/bin/acltool',
+            '/usr/bin/nfs4xdr_winacl',
             '-a', action,
             '-O', str(uid), '-G', str(gid),
-            '-rx' if options['traverse'] else '-r',
+            flags,
             '-c', path,
             '-p', path], check=False, capture_output=True
         )
         if acltool.returncode != 0:
             raise CallError(f"acltool [{action}] on path {path} failed with error: [{acltool.stderr.decode().strip()}]")
 
-    def _common_perm_path_validate(self, path):
-        if not os.path.exists(path):
-            raise CallError(f"Path not found: {path}",
-                            errno.ENOENT)
+    def _common_perm_path_validate(self, schema, path, recursive, verrors):
+        p = Path(path)
+        if not p.is_absolute():
+            verrors.add(f'{schema}.path', 'Must be an absolute path.')
+            return
+
+        if p.is_file() and recursive:
+            verrors.add(f'{schema}.path', 'Recursive operations on a file are invalid.')
+            return
 
         if not os.path.realpath(path).startswith('/mnt/'):
-            raise CallError(f"Changing permissions on paths outside of /mnt is not permitted: {path}",
-                            errno.EPERM)
+            verrors.add(
+                f'{schema}.path',
+                f"Changing permissions on paths outside of /mnt is not permitted: {path}"
```

Comment:
```suggestion
                f"Changing permissions on paths outside of /mnt are not permitted: {path}"
```

Reply:
I don't english real good apparently.

Reply:
Okay. Took me a second to parse what I wrote. In this case the gerund "changing" is the subject of sentence and therefore singular form of verb is correct, but the overall construction is awkwardly phrased.

Reply:
Something something, English are hard...